### PR TITLE
Productionize

### PR DESCRIPTION
**Pull request by callahad**
Still lots of ugly boilerplate to clean up, but I'd like to get this merged and auto-deploying to Heroku.

Would appreciate a review from @stephank or @djc 

Major changes:
1. **Reintroduce Bottle as our web microframework**
   
   Rationale: We're already using PyJWT, so we've accepted higher levels of abstraction. The use of Bottle dramatically simplifies our code and makes it more accessible to developers without intimate knowledge of WSGI. It also brings niceties like compiling caching templates instead of hitting the disk on every request and actually returning 404 errors instead of spewing backtraces when files aren't found.
2. **Use aiohttp, an asyncio based server, instead of wsgiref**
   
   Rationale: We need this thing to survive if it hits the front page of HN. A single-threaded, synchronous server isn't going to cut it. The stdlib has included asyncio since Python 3.4, and aiohttp uses that to great effect.
3. **Allow configuration via environment variables**
   
   Rationale: Effectively necessary for Heroku. See also: https://12factor.net/config
   
   I've also switched to an INI format from JSON so the template can include an explanation of the options.
4. **Use Redis / FakeRedis for nonce storage**
   
   Rationale: This directly mimics what a real RP would need to do, and it works in multi-process environments. It also plugs a memory leak by ensuring that stale, unused nonces get garbage collected.
5.  **Stop keying nonces by email address**
   
   Rationale: It should be possible for the same person to have several simultaneous login attempts, across devices. The previous design precluded that.

I'd like to work further on docs, content, and code hygiene, but I believe this is at a point where we should merge it: it doesn't leak memory, it's deployable on Heroku, it raises fewer exceptions, it's faster, and it's more reliable.

It's currently deployed at https://portier-demo.herokuapp.com/

### CODE REVIEW

File: server.py
Comment (stephank) [inline server.py]: The expiry can be passed as an optional param to `set()`: https://github.com/jamesls/fakeredis/blob/4bfc7a2793aa93723c79512f8be4df2c497498ef/fakeredis.py#L433

Which would also make the transaction unnecessary.

File: requirements.txt
Comment (stephank) [inline requirements.txt]: So this is current master, which is 0.13-dev, I suppose? Looks like we need it for the aiohttp support introduced in [884ef30](https://github.com/bottlepy/bottle/commit/884ef30e1b3fb891783cf56207d2f17f3645de59) (and possible fixes on top of that)?

Could you add a comment here?

**Review by stephank — CHANGES_REQUESTED** [review]
I like all of these changes, and think this is a good direction for code we want to serve as an example. Only a couple of small requests, but perhaps we should also wait for a second review, considering this is the first from me.

Beside the review items, one of the things I think could spawn discussion is the comment style. I personally like having some extra comments in code which is meant to serve as an example, even if it's a bit verbose.

File: server.py
Comment (callahad) [reply server.py]: Even better, Redis has a [SETEX](http://redis.io/commands/setex) command! Fixed in 04463abef85b450810bb450af05b27e68b7cfaa2

File: requirements.txt
Comment (callahad) [reply requirements.txt]: Yep! Done in 4c29a93c7bf35fd8dc11ac8529dbf85601f94695

**Comment by callahad** [comment]
Thanks for the review! Going to work on cleaning up the code and comments real quick.

Also going to hide the configuration stuff in another file; people don't need to see how that sausage gets made. :)

**Comment by callahad** [comment]
Dialed the verbosity to 11. How's it look?

**Comment by callahad** [comment]
I think this is ready to go. It's up and running at https://portier-demo.herokuapp.com/

**Review by stephank — APPROVED** [review]

**Comment by stephank** [comment]
I think these changes are excellent, but I'm also super pragmatic about review. Some of these commits could've been rebased, or contain small unrelated stylistic changes.

So for the record, I feel like a review from @djc would be far more strict. Not sure where we need to be with reviews. :)

**Comment by callahad** [comment]
Given that this is a demo, and not the broker, I'm comfortable merging without going over the commit history with a fine-toothed comb. I would appreciate one final review just to make sure you agree with how I handled #7 in commit 1e31c2488e809038d660daf6e113301c32d17b34

**Comment by callahad** [comment]
ping @stephank -- could I please get one more quick review for commit 1e31c2488e809038d660daf6e113301c32d17b34?

**Comment by stephank** [comment]
@callahad Tested it with some simple curl-fu, and that looks good.

In https://github.com/portier/portier-broker/pull/40 I'm currently returning errors as GET, and I'm not sure which is correct.

**Comment by callahad** [comment]
That's a good point -- you can't easily return them any other way, since you can't cleanly the browser to POST from the backend...

Let's merge without this commit for now and hash out how we're supposed to handle this in discussion on #7.

**Comment by callahad** [comment]
Strike that, I'm rather confident that this implementation is technically correct (per discussion in #7), and thus I'll merge it as-is. We may need to update the broker to return the error data differently.

Whew!

**Pull request merged**
